docs/source/machine_translation.rst

```diff
@@ -0,0 +1,255 @@
+
+Neural machine translation (NMT) models typically operate with a fixed
+vocabulary, but translation is an open-vocabulary problem.
+Byte pair encoding (BPE) enables NMT model translation on open-vocabulary by
+encoding rare and unknown words as sequences of subword units.
+This is based on the intuition that various word classes are translatable via
+smaller units than words. More information in the paper
+https://arxiv.org/abs/1508.07909 BPE creates a list of merges that are used
+for splitting out-of-vocabulary words. Example of such splitting::
+
+  basketball => basket@@ ball
+
+Postprocessing can be manually done by::
+
+  sed "s/@@ //g"
+
+BPE generation
+**************
+
+In order to use BPE, you must first generate merge_file, over all data. This
+file is generated on both source and target dataset.
+You can generate it by running following script::
+
+  neuralmonkey/lib/subword_nmt/learn_bpe.py -s 50000 < DATA > merge_file.bpe
+
+With the data from this tutorial it would be the following command::
+
+  paste Batch1a_en.txt Batch1a_cs.txt |
+  neuralmonkey/lib/subword_nmt/learn_bpe.py -s 8000 >
+  exp-nm-mt/data/merge_file.bpe
+
+You can change number of merges, this number is equivalent to the size of the
+vocabulary. Do not forget that as an input is the file containing both source
+and target sides.
+
+
+
+
+Part II. - The Model Configuration
+----------------------------------
+
+In this section, we will go through setting the configuration file
+``translation.ini`` needed for
+the training of the machine translation task. We will mention only
+differences from the main Post_editing tutorial
+
+1 - Datasets
+************
+
+For training, we prepare two datasets. Since we will be using BPE, we need to
+ define the preprocessor. The configuration of the datasets looks like this::
+
+  [train_data]
+  class=config.utils.dataset_from_files
+  s_source=exp-nm-mt/data/train/Batch1a_en.txt.gz
+  s_target=exp-nm-mt/data/train/Batch1a_cs.txt.gz
+  preprocessor=<bpe_preprocess>
+
+  [val_data]
+  class=config.utils.dataset_from_files
+  s_source=exp-nm-mt/data/dev/Batch2a_en.txt.gz
+  s_target=exp-nm-mt/data/dev/Batch2a_cs.txt.gz
+  preprocessor=<bpe_preprocess>
+
+2 - Preprocessor and postprocessor
+**********************************
+
+Wee need to tell the Neural Monkey how it should handle preprocessing and
+postprocessing due to the BPR::
+
+  [bpe_preprocess]
+  class=processors.bpe.BPEPreprocessor
+  merge_file=exp-nm-mt/data/merge_file.bpe
+
+  [bpe_postprocess]
+  class=processors.bpe.BPEPostprocessor
+
+
+3 - Vocabularies
+****************
+
+For both encoder and decoder we will use shared vocabulary created from BPE
+merges::
+
+  [shared_vocabulary]
+  class=vocabulary.from_bpe
+  path=exp-nm-mt/data/merge_file.bpe
+
+4 - Encoder and Decoder
+************
+
+The encoder and decored is similar to Post-Editing ones::
+
+  [encoder]
+  class=encoders.sentence_encoder.SentenceEncoder
+  name=sentence_encoder
+  rnn_size=300
```

Comment:
Prijde mi hloupost popisovat experiment "ideologicky" kdyz nase opice ho nezvladne na takove urovni ... protoze zere moc musime mit maly modely a dokud neudelame (nebo TF) neco s pametovou narocnosti nemuzem si vymyslet. Takze jsem ochotnej to nastavit tak, aby se to veslo na 12GB GPU ale nesouhlasim davat do dokumentace cisla na imaginarni experiment kterej nikdy nikdo nedokaze spustit.